feat/maximal-repeats-impl

balrajsingh9 opened the pull request:
- Implement maximal repeats
- Update method signatures
- Annotate with comments and types
- Update logic to compute alphabet set(fixes #8)

balrajsingh9 committed:
refactor: add type annotations

balrajsingh9 committed:
fix repeats algorithm and use dict

balrajsingh9 committed:
annotate comments and update method signatures

balrajsingh9 committed:
use bwt table instead of looking up in s and refactor heavily

mailund commented:
You have requested a review on a pull request with the name [Do-Not-Review], so I'm a bit confused. To review or not review, that is the question.

balrajsingh9 commented:
> You have requested a review on a pull request with the name [Do-Not-Review], so I'm a bit confused. To review or not review, that is the question.

Sorry about the ambiguity. I was working on this branch last night and found a lot of bugs and didn't want you to review them before they were fixed by me. But I managed to(probably?) fix them. I have removed the [Do-Not-Review] prefix and changed the PR for ready to review.

balrajsingh9 committed:
chore: update init

mailund commented on core/maximal_repeats.py:
```diff
@@ -20,113 +23,123 @@ class Interval:
+    prev: Optional[Node] = None
+    # init the alphabet set over the input string
+    alphabet_set: set[str] = set(s)
+    # init the dict that maps the alphabet to a pos p, -1 is for undefined at init
```
If you write -1 for undefined, why is the entire alphabet set to zero?

balrajsingh9 replied:
It was a mistake, I have pushed it with -1 now; was experimenting with things and didn't reset it to -1

balrajsingh9 committed:
fix: change back to -1 for undefined in map

mailund commented on core/maximal_repeats.py:
```diff
@@ -20,113 +23,123 @@ class Interval:
+    alphabet_pos_map: dict[str, int] = {alphabet: 0 for alphabet in alphabet_set}
     lcp = lcp_interval.lcp_value
 
     for j in range(0, len(lcp_interval.child_list)):
```
The construction `range(0, len(x))` is considered bad form. First, because `range(0, n)` is the same as `range(n)` and second because of reasons I don't know, but some people have a strong opinion about it. Probably because in most cases where you use `for i in range(len(x))` you don't use `i` anywhere except for in `x[i]` so `for a in x` would be faster and less likely to have bugs. You don't use `j` here, as far as I can tell, except for in `lcp_interal.child_list[j]` so you could have used `for child in lcp_interval.child_list` instead.

balrajsingh9 committed:
refactor: iterate over list without indexing

balrajsingh9 committed:
fix: check if p and p_prime are not -1

balrajsingh9 committed:
refactor: rename the method

balrajsingh9 committed:
fix: don't process the root
